### ISSUE TITLE
multi-echo hangs at the end of the pipeline

### ISSUE THREAD

**Issue by bpinsard** [nipreps/fmriprep#2762]
### What happened?

I tried multiple times to run fMRIPrep on multi-echo data, using the latest singularity image from repronim containers.
It runs for a while, outputs files: dseg and individual TE preproc (if `--me-output-echos` is used) but not the combined preproc data.
It always stops after these confounds steps:

```
...
...
...
220412-17:18:11,369 nipype.workflow INFO:
         [Node] Finished "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.tcompcor".
220412-17:18:13,183 nipype.workflow INFO:
         [Node] Setting-up "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.tcc_metadata_filter" in "/scratch/basile/wdir_hcptrt_me/fmriprep_wf/single_subject_03_wf/func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf/bold_confounds_wf/tcc_metadata_filter".
220412-17:18:13,186 nipype.workflow INFO:
         [Node] Running "tcc_metadata_filter" ("fmriprep.interfaces.confounds.FilterDropped")
220412-17:18:13,197 nipype.workflow INFO:
         [Node] Finished "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.tcc_metadata_filter".
220412-17:18:14,963 nipype.workflow INFO:
         [Node] Setting-up "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.tcc_metadata_fmt" in "/scratch/basile/wdir_hcptrt_me/fmriprep_wf/single_subject_03_wf/func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf/bold_confounds_wf/tcc_metadata_fmt".
220412-17:18:15,178 nipype.workflow INFO:
         [Node] Running "tcc_metadata_fmt" ("niworkflows.interfaces.utility.TSV2JSON")
220412-17:18:15,190 nipype.workflow INFO:
         [Node] Finished "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.tcc_metadata_fmt".
220412-17:18:52,301 nipype.workflow INFO:
         [Node] Finished "fmriprep_wf.single_subject_03_wf.func_preproc_ses_001_task_wm_run_01_echo_1_part_mag_wf.bold_confounds_wf.dvars".
```

No process is using CPU except the main launching process (I guess from polling the workers), the multiprocessing workers are all idle and have not child process.

I don't know if that's a nipype issue or an fMRIPrep one. 

Are there any other info that could help? Ways to probe nipype? 

### What command did you use?

```shell
`singularity run -B /scratch/:/scratch/ -B /data/neuromod:/data/neuromod /data/neuromod/containers/repronim-containers/images/bids/bids-fmriprep--21.0.1.sing -w /scratch/basile/wdir_hcptrt_me --anat-derivatives /data/neuromod/DATA/cneuromod.processed/smriprep --fs-subjects-dir /data/neuromod/DATA/cneuromod.processed/smriprep/sourcedata/freesurfer --output-layout bids --ignore slicetiming --output-spaces MNI152NLin2009cAsym T1w:res-iso2mm --cifti-output 91k --notrack --write-graph --omp-nthreads 8 --nprocs 48 --mem_mb 49152 --fs-license-file /data/neuromod/DATA/cneuromod/code/ds_prep/derivatives/fmriprep/freesurfer.license --bids-filter-file /data/neuromod/DATA/pilots/pilot_processed/fmriprep/hcptrt_seqdev_me/bids_filters.json --me-output-echos /data/neuromod/DATA/pilots/cneuromod/hcptrt_seqdev_me /data/neuromod/DATA/pilots/pilot_processed/fmriprep/hcptrt_seqdev_me participant`
```

### What version of fMRIPrep are you running?

21.0.1

### How are you running fMRIPrep?

Singularity

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

Anatomical derivatives

**Comment by mgxd** [comment]
I'm not sure why this is hanging, but if you run with max verbosity (`-vvv`) you should be able to see all nipype logger messages, including the `MultiProc` status messages. Perhaps that can help?

**Comment by bpinsard** [comment]
Thanks for the pointer, here is the lattest messages that are added by the verbose flags
```
220419-12:38:11,514 nipype.workflow DEBUG:
         Progress: 5271 jobs, 3713/2/0 (done/running/ready), 2/1556 (pending_tasks/waiting).
220419-12:38:11,514 nipype.workflow DEBUG:
         Tasks currently running: 2. Pending: 2.
220419-12:38:11,526 nipype.workflow INFO:
         [MultiProc] Running 2 tasks, and 0 jobs ready. Free memory (GB): 48.75/49.15, Free processors: 46/48.
                     Currently running:
                       * fmriprep_wf.single_subject_03_wf.bids_info
                       * fmriprep_wf.single_subject_01_wf.bids_info
``` 

The 2 running jobs have no corresponding folders in the workdir. 

So I try running one of these 2 nodes alone in the container, giving it the expected inputs.
It turns out that this line is certainly the culprit
https://github.com/nipreps/fmriprep/blob/6ba171f78cc4ece177b68e4953c83bc60c2c9071/fmriprep/workflows/base.py#L304 

`fix_multi_T1w_source_name` is being called on a list of lists as the 3 echoes are grouped in a list.
So that error will only be triggered by multi-echo in combination with precomputed anat derivatives.  Weirdly, this error seems to be silent and the node is still submitted to the MultiProc queue.
Will either require fixing niworkflows function or having a new one in fmriprep that flattens the list.

**Comment by mgxd** [comment]
Nice catch - IMO these in-line connection functions are messy as they can suppress errors, avoid caching, and obfuscate what is truly going on. We should probably move the logic into a `Function` node, either in 21.0.2 or 22.0.0

**Comment by mgxd** [comment]
should be addressed in `21.0.2`
